File: src/main/java/io/vproxy/vfx/ui/table/VTableRowListDelegate.java
package io.vproxy.vfx.ui.table;

import java.util.*;
import java.util.stream.Collectors;

@SuppressWarnings("NullableProblems")
public class VTableRowListDelegate<E> implements List<E> {
    private final List<VTableRow<E>> ls;
    private final VTableSharedData<E> shared;

    public VTableRowListDelegate(List<VTableRow<E>> ls, VTableSharedData<E> shared) {
        this.ls = ls;
        this.shared = shared;
    }

    @Override
    public int size() {
        return ls.size();
    }

    @Override
    public boolean isEmpty() {
        return ls.isEmpty();
    }

    @Override
    public boolean contains(Object o) {
        return ls.stream().anyMatch(e -> Objects.equals(e.item, o));
    }

    @Override
    public Iterator<E> iterator() {
        return new VTableRowIteratorDelegate<>(ls.listIterator(), shared);
    }

    @Override
    public Object[] toArray() {
        return ls.stream().map(s -> s.item).toArray();
    }

    @Override
    public <T> T[] toArray(T[] a) {
        //noinspection SuspiciousToArrayCall
        return ls.stream().map(s -> s.item).collect(Collectors.toList()).toArray(a);
    }

    @Override
    public boolean add(E e) {
        ls.add(new VTableRow<>(e, shared));
        return true;
    }

    @Override
    public boolean remove(Object o) {
        int index = indexOf(o);
        if (index == -1) {
            return false;
        } else {
            ls.remove(index);
            return true;
        }
    }

    @Override
    public boolean containsAll(Collection<?> c) {
        return ls.stream().map(e -> e.item).collect(Collectors.toSet()).containsAll(c);
    }

    @Override
    public boolean addAll(Collection<? extends E> c) {
        var all = new ArrayList<VTableRow<E>>(c.size());
        for (var e : c) {
            all.add(new VTableRow<>(e, shared));
        }
        return ls.addAll(all);
    }

    @Override
    public boolean addAll(int index, Collection<? extends E> c) {
        var all = new ArrayList<VTableRow<E>>(c.size());
        for (var e : c) {
            all.add(new VTableRow<>(e, shared));
        }
        return ls.addAll(index, all);
    }

    @Override
    public boolean removeAll(Collection<?> c) {
        return ls.removeIf(e -> c.contains(e.item));
    }

    @Override
    public boolean retainAll(Collection<?> c) {
        return ls.removeIf(e -> !c.contains(e.item));
    }

    @Override
    public void clear() {
        ls.clear();
    }

    @Override
    public E get(int index) {
        return ls.get(index).item;
    }

    @Override
    public E set(int index, E element) {
        return ls.set(index, new VTableRow<>(element, shared)).item;
    }

    @Override
    public void add(int index, E element) {
        ls.add(index, new VTableRow<>(element, shared));
    }

    @Override
    public E remove(int index) {
        return ls.remove(index).item;
    }

    @Override
    public int indexOf(Object o) {
        int index = -1;
        for (int i = 0; i < ls.size(); i++) {
            var e = ls.get(i);
            if (Objects.equals(e.item, o)) {
                index = i;
                break;
            }
        }
        return index;
    }

    @Override
    public int lastIndexOf(Object o) {
        int index = -1;
        for (int i = ls.size() - 1; i >= 0; --i) {
            var e = ls.get(i);
            if (Objects.equals(e.item, o)) {
                index = i;
                break;
            }
        }
        return index;
    }

    @Override
    public ListIterator<E> listIterator() {
        return new VTableRowIteratorDelegate<>(ls.listIterator(), shared);
    }

    @Override
    public ListIterator<E> listIterator(int index) {
        return new VTableRowIteratorDelegate<>(ls.listIterator(index), shared);
    }

    @Override
    public List<E> subList(int fromIndex, int toIndex) {
        return ls.subList(fromIndex, toIndex).stream().map(e -> e.item).collect(Collectors.toList());
    }
}

File: src-ts/src/ui/table/VTableRowListDelegate.ts
import type { ObservableList } from '../../javafx/Parent.js';
import { VTableRow } from './VTableRow.js';
import { VTableRowIteratorDelegate } from './VTableRowIteratorDelegate.js';
import type { VTableSharedData } from './VTableSharedData.js';

export class VTableRowListDelegate<E> {
  private readonly ls: ObservableList<VTableRow<E>>;
  private readonly shared: VTableSharedData<E>;

  constructor(ls: ObservableList<VTableRow<E>>, shared: VTableSharedData<E>) {
    this.ls = ls;
    this.shared = shared;
  }

  size(): number {
    return this.ls.size();
  }

  isEmpty(): boolean {
    return this.ls.isEmpty();
  }

  contains(o: E): boolean {
    for (const e of this.ls) {
      if (Object.is(e.item, o)) return true;
    }
    return false;
  }

  iterator(): VTableRowIteratorDelegate<E> {
    return new VTableRowIteratorDelegate(this.ls, this.shared);
  }

  toArray(): E[] {
    const arr: E[] = [];
    for (const e of this.ls) arr.push(e.item);
    return arr;
  }

  add(e: E): boolean {
    this.ls.add(new VTableRow<E>(e, this.shared));
    return true;
  }

  remove(o: E): boolean {
    const index = this.indexOf(o);
    if (index === -1) {
      return false;
    }
    this.ls.removeAt(index);
    return true;
  }

  containsAll(c: E[]): boolean {
    const items: E[] = [];
    for (const e of this.ls) items.push(e.item);
    for (const o of c) {
      if (!items.includes(o)) return false;
    }
    return true;
  }

  addAll(c: E[]): boolean {
    const all: VTableRow<E>[] = [];
    for (const e of c) {
      all.push(new VTableRow<E>(e, this.shared));
    }
    this.ls.addAll(...all);
    return true;
  }

  addAllAt(index: number, c: E[]): boolean {
    const all: VTableRow<E>[] = [];
    for (const e of c) {
      all.push(new VTableRow<E>(e, this.shared));
    }
    for (let i = 0; i < all.length; i++) {
      this.ls.insert(index + i, all[i]!);
    }
    return true;
  }

  removeAll(c: E[]): boolean {
    const before = this.ls.size();
    const toRemove: VTableRow<E>[] = [];
    for (const e of this.ls) {
      if (c.includes(e.item)) toRemove.push(e);
    }
    for (const r of toRemove) this.ls.remove(r);
    return this.ls.size() !== before;
  }

  retainAll(c: E[]): boolean {
    const before = this.ls.size();
    const toRemove: VTableRow<E>[] = [];
    for (const e of this.ls) {
      if (!c.includes(e.item)) toRemove.push(e);
    }
    for (const r of toRemove) this.ls.remove(r);
    return this.ls.size() !== before;
  }

  clear(): void {
    this.ls.clear();
  }

  get(index: number): E {
    return this.ls.get(index).item;
  }

  set(index: number, element: E): E {
    const old = this.ls.get(index);
    this.ls.removeAt(index);
    this.ls.insert(index, new VTableRow<E>(element, this.shared));
    return old.item;
  }

  addAt(index: number, element: E): void {
    this.ls.insert(index, new VTableRow<E>(element, this.shared));
  }

  removeAt(index: number): E {
    return this.ls.removeAt(index)!.item;
  }

  indexOf(o: E): number {
    let i = 0;
    for (const e of this.ls) {
      if (Object.is(e.item, o)) return i;
      i++;
    }
    return -1;
  }

  lastIndexOf(o: E): number {
    for (let i = this.ls.size() - 1; i >= 0; i--) {
      const e = this.ls.get(i);
      if (Object.is(e.item, o)) return i;
    }
    return -1;
  }

  listIterator(): VTableRowIteratorDelegate<E> {
    return new VTableRowIteratorDelegate(this.ls, this.shared);
  }

  listIteratorAt(index: number): VTableRowIteratorDelegate<E> {
    return new VTableRowIteratorDelegate(this.ls, this.shared, index);
  }

  subList(fromIndex: number, toIndex: number): E[] {
    const out: E[] = [];
    for (let i = fromIndex; i < toIndex; i++) {
      out.push(this.ls.get(i).item);
    }
    return out;
  }
}
